File: c++/002-AddTwoNumbers.cpp
#include<iostream>
using namespace std;

struct ListNode {
    int val;
    ListNode *next;
    ListNode(int x) : val(x), next(NULL) {}
};

class Solution {
public:
    ListNode *addTwoNumbers(ListNode *l1, ListNode *l2) {
      ListNode *head = NULL, *current = NULL, *last = NULL;
      int x0 = 0;
      while (l1 != NULL || l2 != NULL || x0 != 0) {
        current = new ListNode(x0);
        if (l1 != NULL) current->val += l1->val;
        if (l2 != NULL) current->val += l2->val;
        x0 = current->val / 10;
        current->val %= 10;

        if (last == NULL) {
            head = current; last = current;
        } else {
            last->next = current; last = current;
        }
        
        if (l1 != NULL) l1 = l1->next;
        if (l2 != NULL) l2 = l2->next;
      }    
      return head;
    }
};

int main() {
    return 0;
}

File: python/002-AddTwoNumbers.py
# Definition for singly-linked list.
class ListNode:
     def __init__(self, x):
         self.val = x
         self.next = None

class Solution:
    # @param {ListNode} l1
    # @param {ListNode} l2
    # @return {ListNode}
    def addTwoNumbers(self, l1, l2):
        head, carry, x, y = ListNode(0), 0, 0, 0
        currNode = head        
        while l1 or l2 :
            if l1 :
                x, l1 = l1.val, l1.next
            else :
                x = 0
            if l2 :
                y, l2 = l2.val, l2.next
            else : 
                y = 0 
            s = x + y + carry
            carry = int(s / 10)
            s = s % 10
            currNode.next = ListNode(s)
            currNode = currNode.next
        if carry > 0 :
            currNode.next = ListNode(carry)
        return head.next

l1 = ListNode(3)
l2 = ListNode(9)
solution = Solution()
l = solution.addTwoNumbers(l1, l2)
while l :
    print(l.val)
    l = l.next